### PR TITLE
feat: compile time optimization

### PR DESCRIPTION
**Pull request by NishantJoshi00**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [x] Enhancement
- [ ] Refactoring
- [ ] Dependency updates

## Description

This PR precisely has 2 changes.
1. It removes unnecessary features from diesel
2. It puts `vergen` under feature gate `(to be reviewed and discussed if necessary)`

The corresponding benchmarks are listed in below mentioned comments.
> generated using `cargo r --timings`

### Additional Changes

- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Comment by NishantJoshi00** [comment]
### Benchmarks
> No optimization
Screenshot 2023-03-20 at 10 39 05 PM

**Comment by NishantJoshi00** [comment]
### Benchmarks
> `Diesel feature removal` + `vergen feature gating`
Screenshot 2023-03-20 at 10 41 22 PM

> PS: considering this is not a significant change, this can be brought back, while still keeping decent performance. (decision pending)
> Identified bottleneck (libgit2-sys build script)

**Review by jarnura — COMMENTED** [review]
currently the functionality is dropped for `generate_cargo_instructions`, adding it as a feature flag is fine.

**Review by jarnura — APPROVED** [review]

**Review by dracarys18 — DISMISSED** [review]

File: crates/drainer/src/settings.rs [hunk 1]
```diff
@@ -9,6 +9,16 @@ use serde::Deserialize;
...
+pub struct CmdLineConf {
+    /// Config file.
+    /// Application will look for "config/config.toml" if this option isn't specified.
+    #[arg(short = 'f', long, value_name = "FILE")]
+    pub config_path: Option<PathBuf>,
+}
+
+#[cfg(feature = "vergen")]
 #[derive(clap::Parser, Default)]
 #[command(version = router_env::version!())]
```
Comment (SanchithHegde) [inline crates/drainer/src/settings.rs:23]: Add this part behind `cfg_attr`?

```suggestion
#[cfg_attr(feature = "vergen", command(version = router_env::version!()))]
```

File: crates/router/src/configs/settings.rs [hunk 2]
```diff
@@ -15,6 +15,19 @@ use crate::{
...
+    #[arg(short = 'f', long, value_name = "FILE")]
+    pub config_path: Option<PathBuf>,
+
+    #[command(subcommand)]
+    pub subcommand: Option<Subcommand>,
+}
+
+#[cfg(feature = "vergen")]
 #[derive(clap::Parser, Default)]
 #[command(version = router_env::version!())]
```
Comment (SanchithHegde) [inline crates/router/src/configs/settings.rs:32]:
```suggestion
#[cfg_attr(feature = "vergen", command(version = router_env::version!()))]
```

File: crates/router_env/Cargo.toml [hunk 3]
```diff
@@ -33,10 +33,11 @@ vergen = { version = "8.0.0-beta.9", optional = true, features = ["cargo", "git"
 tokio = { version = "1.26.0", features = ["macros", "rt-multi-thread"] }
 
 [build-dependencies]
-vergen = { version = "8.0.0-beta.9", features = ["cargo", "git", "git2", "rustc"] }
+vergen = { version = "8.0.0-beta.9", features = ["cargo", "git", "git2", "rustc"], optional = true }
 
 [features]
 default = ["actix_web"]
 actix_web = ["tracing-actix-web"]
+# vergen = ["dep:vergen"]
```
Comment (SanchithHegde) [inline crates/router_env/Cargo.toml:41]: Please remove this 😬

**Review by jarnura — APPROVED** [review]

**Pull request merged**